app/code/Magento/CatalogCustomerGraphQl/Model/Resolver/PriceTiers.php

```diff
@@ -120,42 +120,89 @@ function () use ($productId, $context) {
-
-                return $this->formatProductTierPrices($tierPrices, $productPrice, $store);
+                return $this->formatAndFilterTierPrices($tierPrices, $productPrice, $store);
             }
         );
     }
 
     /**
-     * Format tier prices for output
+     * Format and filter tier prices for output
      *
      * @param ProductTierPriceInterface[] $tierPrices
      * @param float $productPrice
      * @param StoreInterface $store
      * @return array
      */
-    private function formatProductTierPrices(array $tierPrices, float $productPrice, StoreInterface $store): array
-    {
+    private function formatAndFilterTierPrices(
+        array $tierPrices,
+        float $productPrice,
+        StoreInterface $store
+    ): array {
         $tiers = [];
+        $qtyCache = [];
+
+        foreach ($tierPrices as $key => $tierPrice) {
+            $this->formatTierPrices($productPrice, $store, $tierPrice, $tiers);
+            $this->filterTierPrices($tierPrices, $key, $tierPrice, $qtyCache, $tiers);
+        }
+        return $tiers;
+    }
+
+    /**
+     * Format tier prices for output
+     *
+     * @param float $productPrice
+     * @param StoreInterface $store
+     * @param ProductTierPriceInterface $tierPrice
+     * @param array $tiers
+     */
+    private function formatTierPrices(float $productPrice, StoreInterface $store, &$tierPrice, &$tiers)
+    {
+        $tierPrice->setValue($this->priceCurrency->convertAndRound($tierPrice->getValue()));
+        $percentValue = $tierPrice->getExtensionAttributes()->getPercentageValue();
+        if ($percentValue && is_numeric($percentValue)) {
+            $discount = $this->discount->getDiscountByPercent($productPrice, (float)$percentValue);
+        } else {
+            $discount = $this->discount->getDiscountByDifference($productPrice, (float)$tierPrice->getValue());
+        }
 
-        foreach ($tierPrices as $tierPrice) {
-            $tierPrice->setValue($this->priceCurrency->convertAndRound($tierPrice->getValue()));
-            $percentValue = $tierPrice->getExtensionAttributes()->getPercentageValue();
-            if ($percentValue && is_numeric($percentValue)) {
-                $discount = $this->discount->getDiscountByPercent($productPrice, (float)$percentValue);
+        $tiers[] = [
+            "discount" => $discount,
+            "quantity" => $tierPrice->getQty(),
+            "final_price" => [
+                "value" => $tierPrice->getValue(),
+                "currency" => $store->getCurrentCurrencyCode()
+            ]
+        ];
+    }
+
+    /**
+     * Filter the lowest price for each quantity
+     *
+     * @param array $tierPrices
+     * @param int $key
+     * @param ProductTierPriceInterface $tierPriceItem
+     * @param array $qtyCache
+     * @param array $tiers
+     */
+    private function filterTierPrices(
+        array $tierPrices,
+        int $key,
+        ProductTierPriceInterface $tierPriceItem,
+        array &$qtyCache,
```

Comment:
$qtyCache doesn't need to be passed by reference
it can be part of your object \Magento\CatalogCustomerGraphQl\Model\Resolver\PriceTiers
by declaring a private $productTiersByQuanity[] //you can dcument that this acts like a cache
it can hold prices like $productTiersByQuanity[customerGroup][productId][qty] so it can be used for multiple products and if called multiple times for the same product, you can just fetch it and not even need to recalculate it.
not sure if customerGroup is required. probably not. you'l only get the relevant tiers for current customerGroup
of course filterTierPrices : array which is the filtered tiers, and you can pass formatted tiers from previous function return

Reply:
**Done**